### DISCUSSION TITLE
Auto Materialization: No conditions met when it should have been

### DISCUSSION THREAD

**Discussion by MohamedMeguid** [dagster-io/dagster#18359] [Q&A] [answered]
Hello!

We have a PoC of Dagster for managing dbt with ~450 assets deployed on Kubernetes where we observed that assets aren't being materialized according to the specified freshness policy, checking the Auto Materialization history, we see many evaluations with no conditions met but when we checked manually all the upstream models, they are materialized on time.

Is there something we can do to make sure assets are triggered according to the freshness policies.

The following screenshots show an example:

![Screenshot from 2023-11-29 11-10-29](https://github.com/dagster-io/dagster/assets/18240811/c616d7b0-c8a9-40d1-b93d-40516676ace1)
![Screenshot from 2023-11-29 11-10-29-2](https://github.com/dagster-io/dagster/assets/18240811/26340150-4b59-40ee-8777-165d2676dbc7)
_The last skip was due to "Waiting on upstream data to be up to date", which were up to date just a minute after this tick._

Thanks!

**Comment by MohamedMeguid** [comment]
I think this is related to this [issue](https://github.com/dagster-io/dagster/issues/18042), that the **No conditions met** title is misleading and it is rather nothing changed since the last skipped tick.

For the sake of having the asset of interest overdue by a few minutes rather that overdue by hours, I have disabled the `skip_on_parent_outdated` rule and the assets started to materialize, seems there was always an asset in the upstream that was always a minute or two late so the asset was always skipped.

But still want to discuss if I need to make sure my assets adhere to the freshness policy and the lineage is somewhat big and complex (100+ upstream models), what can I do?

**Comment by OwenKephart** [comment]
Hi @MohamedMeguid -- I think another related discussion would be this one: https://github.com/dagster-io/dagster/issues/18325#issuecomment-1830485890. The short answer is that the freshness calculation is very "global" and in many cases is pretty overkill compared to what you actually need, and generally harder to predict than the other rules, which is a reason we're moving away from that API for most use cases.

The best solution here depends on the specific semantics you need for the downstream assets. My assumption is that, at a high level, you just want your downstream assets to update less frequently than the upstreams (otherwise you could just use the eager policy).

One simpler option here would be something along the lines of "materialize this asset once every hour, after all of its parents have updated". This would ensure that:

a) this asset will update once per X time period (rather than every time a parent is updated)
b) this asset will wait for its parents to be updated before it gets updated itself (so it won't do any "useless" updates)

To create a policy like this, you could do:

```python
my_policy = AutoMaterializePolicy.eager().without_rules(
    AutoMaterializeRule.materialize_on_parent_updated(),
).with_rules(
    AutoMaterializeRule.materialize_on_cron("<your cron string>"),
    AutoMaterializeRule.skip_on_not_all_parents_updated(),
)
```

This set of APIs is something we're in the process of reworking to make these sorts of tasks simpler, so I'm also curious in learning what the ideal behavior you're looking for is here.

**Comment by MohamedMeguid** [comment]
Hi @OwenKephart, Thank you for your comment, it's very helpful!

Our use case is as following, we have a set of core tables (e.g. dim_customers, fct_orders tables) which depend on more than 100+ other assets over a series of layer, for example:
`[base_customers_1, base_customers_2, ...] -> stg_customers -> int_customers -> dim_customers`
`[base_orders_1, base_orders_2, ...] -> stg_orders -> int_orders -> fct_orders -> dim_customers`
`[stg_products_src_1, stg_products_src_2] -> int_products -> dim_products -> fct_orders`
and so on.

We need the `fct_orders` to be 30 minutes fresh dictated by business requirements as it is used for operational reasons, other downstream core tables like the `dim_customers` isn't needed as frequently. And the `fct_orders` has so many upstream deps that it isn't convenient to specify a cron schedule for all of them. So we thought the Freshness Policy here made sense.

What can help here is that if we can specify a `AutoMaterializeRule.materialize_on_cron("0,30 * * * *")` for the `fct_orders`, and another `AutoMaterializeRule.materialize_on_cron("0 * * * *")` for the `dim_customers`  and have the schedules propagate upstream, so that each upstream asset materializes the most frequent of all schedules (I think this is exactly the idea of the Freshness Policy).

As a summary, we think managing schedules from terminal nodes and assets of business interest makes more sense than managing root nodes that has so many intricacies.

**Reply by OwenKephart** [reply, marked as answer]
@MohamedMeguid Got it, that makes sense to me, and I definitely see how the terminal node approach is easier for you here. The tricky bit is just the number of unknowns in that calculation -- it's hard to estimate how long the upstream nodes will take to run, and so kicking things off at the correct time is fairly error-prone (and so we've invested more in more explicit ways of specifying when particular assets should run).

One (potentially wacky) idea would be to programmatically do this "propagation" process. If you happen to be using dbt here, you should be able to operate over the manifest and apply a specific automaterialize policy to anything upstream of `fct_orders` using the DagsterDbtTranslator.

Otherwise, it is possible to get Dagster's representation of the graph of assets from your Definitions object (`defs.get_asset_graph()`) so there is conceivably a way to use that to create a new Definitions object with your desired AMPs applied.

Your use case is definitely valid, this is just one of those areas where the fully-general solution will take some time to get right, so in the meantime workarounds are probably the best bet.

**Comment by MohamedMeguid** [comment]
@OwenKephart Cool! Thank you, Owen!
